compiler/src/iree/compiler/Codegen/Common/GPU/GPUConvertToCoalescedDMA.cpp

```diff
@@ -33,223 +36,444 @@ namespace mlir::iree_compiler {
-  if (indicesType.getElementType().isIndex()) {
-    return indices;
+/// Create a coalesced DMA operation in the in_parallel region.
+/// Handles both copy and gather operations.
+template <typename OpTy>
+static LogicalResult createDMAInForall(scf::ForallOp threadForallOp,
+                                       PatternRewriter &rewriter) {
+  // Find the inner operation.
+  OpTy innerOp = nullptr;
+  threadForallOp->walk([&](OpTy foundOp) {
+    innerOp = foundOp;
+    return WalkResult::interrupt();
+  });
+
+  if (!innerOp) {
+    return failure();
+  }
+
+  Block *forallBody = threadForallOp.getBody();
+  Value sharedOut = forallBody->getArguments().back();
+  size_t numIVs = forallBody->getNumArguments() - 1;
+  Value laneId = forallBody->getArgument(numIVs - 1);
+
+  auto inParallelOp = cast<scf::InParallelOp>(forallBody->getTerminator());
+  Block &inParallelBlock = inParallelOp.getRegion().front();
+
+  // Collect parallel_insert_slice ops to erase.
+  SmallVector<tensor::ParallelInsertSliceOp> toErase;
+  for (auto &op : inParallelBlock) {
+    if (auto insertOp = dyn_cast<tensor::ParallelInsertSliceOp>(&op)) {
+      toErase.push_back(insertOp);
+    }
+  }
+
+  Location loc = innerOp.getLoc();
+  Value source, indices;
+
+  // Extract source and indices based on op type.
+  if constexpr (std::is_same_v<OpTy, linalg::CopyOp>) {
+    Value input = innerOp.getInputs()[0];
+    if (auto extractSlice = input.getDefiningOp<tensor::ExtractSliceOp>()) {
+      source = extractSlice.getSource();
+    } else {
+      return failure();
+    }
+  } else if constexpr (std::is_same_v<OpTy, IREE::LinalgExt::GatherOp>) {
+    source = innerOp.getSource();
+    indices = innerOp.getIndices();
+
+    // Convert indices tensor to vector for DMA if present.
+    if (indices) {
+      rewriter.setInsertionPoint(inParallelOp);
+      auto indicesType = cast<RankedTensorType>(indices.getType());
+      Type elementType = indicesType.getElementType();
+
+      // First, read the indices tensor as a vector with the original element
+      // type
+      VectorType vectorTypeOriginal =
+          VectorType::get(indicesType.getShape(), elementType);
+
+      SmallVector<Value> readIndices(indicesType.getRank());
+      for (int64_t i = 0; i < indicesType.getRank(); ++i) {
+        readIndices[i] = arith::ConstantIndexOp::create(rewriter, loc, 0);
+      }
+
+      // Create padding value - use i32 for index type
+      Type paddingType = elementType;
+      if (elementType.isIndex()) {
+        paddingType = rewriter.getI32Type();
+      }
+      TypedAttr zeroPadAttr = rewriter.getIntegerAttr(paddingType, 0);
+      Value zeroPad = arith::ConstantOp::create(rewriter, loc, zeroPadAttr);
+
+      Value indicesVec = vector::TransferReadOp::create(
+          rewriter, loc, vectorTypeOriginal, indices, readIndices, zeroPad);
+
+      // Convert to i32 type if needed
+      Type i32Type = rewriter.getI32Type();
+      if (elementType != i32Type) {
```

Comment:
Strong disagree with using `i32` here?

Indices are `index` typed - we can `i32`ify them during lowering

Reply:
@krzysz00 Hit the issue when trying to convert from linalg_ext.gather, where its indices are i32 type.

Still want to keep index type here ?